File: src/lidar.rs
use cassini::process_single_tile_lidar_step;
use log::info;
use reqwest::blocking::Client;
use std::time::Instant;
use std::{fs::create_dir_all, path::Path};

use crate::utils::{compress_directory, download_file, upload_file};

pub fn lidar_step(
    tile_id: &str,
    laz_file_url: &str,
    worker_id: &str,
    token: &str,
    base_api_url: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let lidar_files_path = Path::new("lidar-files");
    let lidar_file_path = lidar_files_path.join(format!("{}.laz", &tile_id));

    if !lidar_files_path.exists() {
        create_dir_all(lidar_files_path)?;
    }

    info!("Downloading laz file for tile {}", &tile_id);
    let start = Instant::now();
    let client = Client::new();
    download_file(&client, &laz_file_url, &lidar_file_path, None)?;
    let duration = start.elapsed();

    info!("Laz file for tile {} downloaded in {:.1?}", &tile_id, duration);

    let lidar_step_path = Path::new("lidar-step");

    if !lidar_step_path.exists() {
        create_dir_all(lidar_step_path)?;
    }

    let output_dir_path = lidar_step_path.join(&tile_id);

    info!("Processing LiDAR step for tile {}", &tile_id);
    let start = Instant::now();

    process_single_tile_lidar_step(&lidar_file_path, &output_dir_path);

    let duration = start.elapsed();

    info!("LiDAR step for tile {} processed in {:.1?}", &tile_id, duration);

    info!("Compressing resulting files for tile {}", &tile_id);
    let start = Instant::now();

    let archive_file_name = format!("{}.tar.xz", &tile_id);
    let archive_path = lidar_step_path.join(&archive_file_name);
    compress_directory(&output_dir_path, &archive_path)?;

    let duration = start.elapsed();

    info!(
        "Resulting files compression for tile {} done in {:.1?}",
        &tile_id, duration
    );

    let url = format!("{}/api/map-generation/lidar-steps/{}", base_api_url, &tile_id);

    upload_file(
        &client,
        worker_id,
        token,
        url,
        base_api_url,
        archive_file_name,
        archive_path,
        "application/x-bzip2",
    )?;

    Ok(())
}

File: src/lidar.ts
import { ensureDir, exists } from "@std/fs";
import { join } from "@std/path";
import {
  LIDAR_FILES_DIR_NAME,
  LIDAR_STEP_DIR_NAME,
  LIDAR_STEP_ENDPOINT_PATH,
} from "./constants.ts";
import { LidarJob } from "./next-job-schema.ts";
import { compressDirectory, executeCommand, log } from "./utils.ts";
import { JobHandlingAdditionnalArguments } from "./models.ts";

export async function handleLidarJob(
  { tileId, tileUrl }: LidarJob["data"],
  { threadNumber, mapantApiWorkerId, mapantApiToken, mapantApiBaseUrl }:
    JobHandlingAdditionnalArguments,
) {
  const lidarFilePath = join(LIDAR_FILES_DIR_NAME, `${tileId}.laz`);
  const lidarStepOutputDirPath = join(LIDAR_STEP_DIR_NAME, tileId);
  const lidarStepArchiveFileName = `${tileId}.tar.xz`;
  const lidarStepArchivePath = join(LIDAR_STEP_DIR_NAME, lidarStepArchiveFileName);

  try {
    await ensureDir(LIDAR_FILES_DIR_NAME);

    log(`Tile ${tileId} | Downloading LiDAR file`, { level: "info", threadNumber });
    const fileResponse = await fetch(tileUrl);

    if (!fileResponse.ok || !fileResponse.body) {
      throw new Error(
        `Could not download LiDAR file.\nResponse status: ${fileResponse.status}.\n${await fileResponse
          .text()}`,
      );
    }

    const file = await Deno.open(lidarFilePath, { write: true, create: true });
    await fileResponse.body.pipeTo(file.writable);

    log(`Tile ${tileId} | LiDAR file downloaded`, { level: "info", threadNumber });

    await ensureDir(lidarStepOutputDirPath);

    log(`Tile ${tileId} | Running cassini lidar step`, { level: "info", threadNumber });
    await executeCommand("cassini", "lidar", lidarFilePath, "-o", lidarStepOutputDirPath);
    await checkIfCassiniExecutionWentOk(lidarStepOutputDirPath);
    log(`Tile ${tileId} | Cassini lidar step done`, { level: "info", threadNumber });

    log(`Tile ${tileId} | Compressing lidar step`, { level: "info", threadNumber });
    await compressDirectory(lidarStepOutputDirPath, lidarStepArchivePath);
    log(`Tile ${tileId} | Compressing lidar step done`, { level: "info", threadNumber });

    log(`Tile ${tileId} | Uploading lidar step`, { level: "info", threadNumber });
    const archiveFile = await Deno.readFile(lidarStepArchivePath);
    const formData = new FormData();

    formData.append(
      "file",
      new Blob([archiveFile], { type: "application/x-bzip2" }),
      lidarStepArchiveFileName,
    );

    const uploadResponse = await fetch(`${mapantApiBaseUrl}${LIDAR_STEP_ENDPOINT_PATH}/${tileId}`, {
      method: "POST",
      body: formData,
      headers: {
        "Origin": mapantApiBaseUrl,
        "Authorization": `Bearer ${mapantApiWorkerId}.${mapantApiToken}`,
      },
    });

    if (!uploadResponse.ok) {
      throw new Error(
        `Could not upload LiDAR step result.\nResponse status: ${uploadResponse.status}.\n${await uploadResponse
          .text()}`,
      );
    }

    log(`Tile ${tileId} | Lidar step result upload done`, { level: "info", threadNumber });
  } catch (e) {
    if (await exists(lidarFilePath)) Deno.remove(lidarFilePath);
    if (await exists(lidarStepArchivePath)) Deno.remove(lidarStepArchivePath);

    if (await exists(lidarStepOutputDirPath)) {
      Deno.remove(lidarStepOutputDirPath, { recursive: true });
    }

    throw e;
  }

  Deno.remove(lidarFilePath);
  Deno.remove(lidarStepArchivePath);
}

async function checkIfCassiniExecutionWentOk(lidarStepOutputDirPath: string) {
  if (
    !(await exists(join(lidarStepOutputDirPath, "dem.tif"))) ||
    !(await exists(join(lidarStepOutputDirPath, "dem-low-resolution.tif"))) ||
    !(await exists(join(lidarStepOutputDirPath, "high-vegetation.tif"))) ||
    !(await exists(join(lidarStepOutputDirPath, "medium-vegetation.tif"))) ||
    !(await exists(join(lidarStepOutputDirPath, "extent.txt"))) ||
    !(await exists(join(lidarStepOutputDirPath, "pipeline.json")))
  ) {
    throw new Error("An error occured while executing cassini lidar command");
  }
}
